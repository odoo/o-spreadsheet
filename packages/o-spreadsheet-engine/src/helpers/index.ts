export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
