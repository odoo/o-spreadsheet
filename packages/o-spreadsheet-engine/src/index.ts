export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
