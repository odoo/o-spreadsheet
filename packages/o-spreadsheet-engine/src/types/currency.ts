export interface Currency {
  name: string;
  code: string;
  symbol: string;
  decimalPlaces: number;
  position: "before" | "after";
}
