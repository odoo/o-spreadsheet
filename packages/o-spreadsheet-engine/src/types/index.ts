export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
