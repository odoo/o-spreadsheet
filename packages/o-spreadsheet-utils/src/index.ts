export * from "./translation";
export * from "./uuid";
