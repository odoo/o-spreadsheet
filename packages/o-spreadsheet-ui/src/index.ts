import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}