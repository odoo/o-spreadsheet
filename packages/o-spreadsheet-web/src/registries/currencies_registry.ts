import { Currency } from "../types";
import { Registry } from "./registry";

/**
 * Registry intended to support usual currencies. It is mainly used to create
 * currency formats that can be selected or modified when customizing formats.
 */
export const currenciesRegistry = new Registry<Currency>();
