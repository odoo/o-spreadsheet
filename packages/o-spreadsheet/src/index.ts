import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };