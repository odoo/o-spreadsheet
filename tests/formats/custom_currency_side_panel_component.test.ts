import { Model } from "../../src";
import { CustomCurrencyPanel } from "../../src/components/side_panel/custom_currency/custom_currency";
import { currenciesRegistry } from "../../src/registries/currencies_registry";
import { Currency } from "../../src/types/currency";
import { setSelection, updateLocale } from "../test_helpers/commands_helpers";
import { FR_LOCALE } from "../test_helpers/constants";
import { click, setInputValueAndTrigger } from "../test_helpers/dom_helper";
import { mountComponent, nextTick, spyModelDispatch } from "../test_helpers/helpers";
jest.mock("../../src/helpers/uuid", () => require("../__mocks__/uuid"));
jest.useFakeTimers();

const selectors = {
  availableCurrencies: ".o-custom-currency .o-available-currencies",
  inputCode: ".o-custom-currency .o-subsection-left input",
  inputSymbol: ".o-custom-currency .o-subsection-right input",
  formatProposals: ".o-custom-currency .o-format-proposals",
  formatProposalOptions: ".o-custom-currency .o-format-proposals option",
  applyFormat: ".o-custom-currency .o-sidePanelButtons button",
};

const [code1, code2] = ["ABC", "DEF"];
const [symbol1, symbol2] = ["µ", "XD"];
const [decimalPlaces1, decimalPlaces2] = [3, 4];
const [position1, position2]: ("after" | "before")[] = ["after", "before"];

const currenciesData: Currency[] = [
  {
    name: "3 decimal places / expression position after",
    code: code1,
    symbol: symbol1,
    decimalPlaces: decimalPlaces1,
    position: position1,
  },
  {
    name: "4 decimal places / expression position before",
    code: code2,
    symbol: symbol2,
    decimalPlaces: decimalPlaces2,
    position: position2,
  },
];

const loadCurrencies = async () => {
  return currenciesData;
};

let dispatch: jest.SpyInstance;
let currenciesContent: { [key: string]: Currency };
let model: Model;
let fixture: HTMLElement;

describe("custom currency sidePanel component", () => {
  beforeEach(async () => {
    currenciesContent = Object.assign({}, currenciesRegistry.content);

    ({ model, fixture } = await mountComponent(CustomCurrencyPanel, {
      env: { loadCurrencies },
      model: new Model({}, { external: { loadCurrencies } }),
      props: { onCloseSidePanel: () => {} },
    }));
    dispatch = spyModelDispatch(model);
    await nextTick();
  });

  afterEach(() => {
    currenciesRegistry.content = currenciesContent;
  });

  describe("Sidepanel", () => {
    // -------------------------------------------------------------------------
    // available currencies selector
    // -------------------------------------------------------------------------

    test("select currency in available currencies selector --> changes code input and symbol input", async () => {
      expect((document.querySelector(selectors.inputCode) as HTMLInputElement).value).toBe("");
      expect((document.querySelector(selectors.inputSymbol) as HTMLInputElement).value).toBe("");

      await setInputValueAndTrigger(selectors.availableCurrencies, "1");

      expect((document.querySelector(selectors.inputCode) as HTMLInputElement).value).toBe(code1);
      expect((document.querySelector(selectors.inputSymbol) as HTMLInputElement).value).toBe(
        symbol1
      );
    });

    test("select currency in available currencies selector --> changes currency proposals", async () => {
      await setInputValueAndTrigger(selectors.availableCurrencies, "1");
      expect(document.querySelector(selectors.formatProposals)).toMatchSnapshot();

      await setInputValueAndTrigger(selectors.availableCurrencies, "2");
      expect(document.querySelector(selectors.formatProposals)).toMatchSnapshot();
    });

    test("select currency in available currencies selector --> does not change proposal selected index", async () => {
      await setInputValueAndTrigger(selectors.availableCurrencies, "1");
      await setInputValueAndTrigger(selectors.formatProposals, "6");
      expect((document.querySelector(selectors.formatProposals) as HTMLSelectElement).value).toBe(
        "6"
      );

      await setInputValueAndTrigger(selectors.availableCurrencies, "2");
      expect((document.querySelector(selectors.formatProposals) as HTMLSelectElement).value).toBe(
        "6"
      );
    });

    test("select first currency in available currencies selector --> remove code input and symbol input", async () => {
      await setInputValueAndTrigger(selectors.availableCurrencies, "1");

      expect((document.querySelector(selectors.inputCode) as HTMLInputElement).value).toBe(code1);
      expect((document.querySelector(selectors.inputSymbol) as HTMLInputElement).value).toBe(
        symbol1
      );

      await setInputValueAndTrigger(selectors.availableCurrencies, "0");

      expect((document.querySelector(selectors.inputCode) as HTMLInputElement).value).toBe("");
      expect((document.querySelector(selectors.inputSymbol) as HTMLInputElement).value).toBe("");
    });

    // -------------------------------------------------------------------------
    // Input Symbol and Input Code
    // -------------------------------------------------------------------------

    test("disable formatProposals/applyFormat if inputSymbol and inputCode are empty", async () => {
      setInputValueAndTrigger(selectors.inputSymbol, "$");
      await setInputValueAndTrigger(selectors.inputCode, "");
      expect(
        (document.querySelector(selectors.formatProposals) as HTMLSelectElement).disabled
      ).toBe(false);
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        false
      );

      setInputValueAndTrigger(selectors.inputSymbol, "");
      await setInputValueAndTrigger(selectors.inputCode, "USD");
      expect(
        (document.querySelector(selectors.formatProposals) as HTMLSelectElement).disabled
      ).toBe(false);
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        false
      );

      setInputValueAndTrigger(selectors.inputSymbol, "  ");
      await setInputValueAndTrigger(selectors.inputCode, "");
      expect(
        (document.querySelector(selectors.formatProposals) as HTMLSelectElement).disabled
      ).toBe(true);
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );
    });

    test("disable formatProposals/applyFormat if selected cells formats are same that the select currency format", async () => {
      setSelection(model, ["A1", "A2"]);
      await nextTick();
      await setInputValueAndTrigger(selectors.inputSymbol, "$");
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        false
      );
      await click(fixture, selectors.applyFormat);
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );

      setSelection(model, ["A2"]);
      await nextTick();
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );
      await setInputValueAndTrigger(selectors.inputSymbol, "€");
      await click(fixture, selectors.applyFormat);
      setSelection(model, ["A1", "A2"]);

      await nextTick();
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        false
      );
    });

    test("not disable formatProposals/applyFormat if apply proposal and select other proposal", async () => {
      setInputValueAndTrigger(selectors.inputSymbol, "$");
      await setInputValueAndTrigger(selectors.inputCode, "USD");
      await setInputValueAndTrigger(selectors.formatProposals, "1");
      await click(fixture, selectors.applyFormat);
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );

      await setInputValueAndTrigger(selectors.formatProposals, "2");
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        false
      );

      await setInputValueAndTrigger(selectors.formatProposals, "1");
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );
    });

    test("not disable formatProposals/applyFormat if apply proposal and select other currency", async () => {
      await setInputValueAndTrigger(selectors.availableCurrencies, "1");
      await click(fixture, selectors.applyFormat);
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );

      await setInputValueAndTrigger(selectors.availableCurrencies, "2");
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        false
      );

      await setInputValueAndTrigger(selectors.availableCurrencies, "1");
      expect((document.querySelector(selectors.applyFormat) as HTMLButtonElement).disabled).toBe(
        true
      );
    });

    test.each([selectors.inputCode, selectors.inputSymbol])(
      "change code input or symbol input --> init available currencies",
      async (selector) => {
        await setInputValueAndTrigger(selectors.availableCurrencies, "1");
        expect(
          (document.querySelector(selectors.availableCurrencies) as HTMLSelectElement).value
        ).toBe("1");

        await setInputValueAndTrigger(selector, "test");
        expect(
          (document.querySelector(selectors.availableCurrencies) as HTMLSelectElement).value
        ).toBe("0");
      }
    );

    test.each([selectors.inputCode, selectors.inputSymbol])(
      "change code input or symbol input --> change currency proposals",
      async (selector) => {
        await setInputValueAndTrigger(selectors.inputCode, "CODE");
        await setInputValueAndTrigger(selectors.inputSymbol, "SYMBOL");
        expect(document.querySelector(selectors.formatProposals)).toMatchSnapshot();

        await setInputValueAndTrigger(selector, "TEST");
        expect(document.querySelector(selectors.formatProposals)).toMatchSnapshot();
      }
    );

    test("currency proposals uses locale format", async () => {
      await setInputValueAndTrigger(selectors.availableCurrencies, "1");
      const proposals = [...document.querySelectorAll(selectors.formatProposalOptions)];
      expect(proposals.map((el) => el.textContent)).toEqual([
        "1,000µ",
        "1,000.000µ",
        "1,000 ABC µ",
        "1,000.000 ABC µ",
        "µ1,000",
        "µ1,000.000",
        "ABC µ1,000",
        "ABC µ1,000.000",
      ]);
      updateLocale(model, FR_LOCALE);
      await nextTick();
      expect(proposals.map((el) => el.textContent)).toEqual([
        "1 000µ",
        "1 000,000µ",
        "1 000 ABC µ",
        "1 000,000 ABC µ",
        "µ1 000",
        "µ1 000,000",
        "ABC µ1 000",
        "ABC µ1 000,000",
      ]);
    });

    test.each([selectors.inputCode, selectors.inputSymbol])(
      "change code input or symbol input --> does not change proposal selected index",
      async (selector) => {
        await setInputValueAndTrigger(selectors.availableCurrencies, "1");
        await setInputValueAndTrigger(selectors.formatProposals, "6");
        expect((document.querySelector(selectors.formatProposals) as HTMLSelectElement).value).toBe(
          "6"
        );

        await setInputValueAndTrigger(selector, "TEST");
        expect((document.querySelector(selectors.formatProposals) as HTMLSelectElement).value).toBe(
          "6"
        );
      }
    );

    test.each([selectors.inputCode, selectors.inputSymbol])(
      "have only one input filled --> display 4 proposals instead of 8",
      async (selector) => {
        await setInputValueAndTrigger(selectors.availableCurrencies, "1");
        expect(document.querySelectorAll(selectors.formatProposals + " option").length).toBe(8);

        await setInputValueAndTrigger(selector, "  ");

        expect(document.querySelectorAll(selectors.formatProposals + " option").length).toBe(4);
      }
    );

    // -------------------------------------------------------------------------
    // Currency proposals
    // -------------------------------------------------------------------------
    test.each([
      ["0", "#,##0[$$]"],
      ["1", "#,##0.00[$$]"],
      ["2", "#,##0[$ USD $]"],
      ["3", "#,##0.00[$ USD $]"],
      ["4", "[$$]#,##0"],
      ["5", "[$$]#,##0.00"],
      ["6", "[$USD $]#,##0"],
      ["7", "[$USD $]#,##0.00"],
    ])("format applied depends on selected proposal", async (proposalIndex, formatResult) => {
      await setInputValueAndTrigger(selectors.inputSymbol, "$");
      await setInputValueAndTrigger(selectors.inputCode, "USD");
      await setInputValueAndTrigger(selectors.formatProposals, proposalIndex);
      await click(fixture, selectors.applyFormat);
      expect(dispatch).toHaveBeenCalledWith("SET_FORMATTING", {
        sheetId: model.getters.getActiveSheetId(),
        target: model.getters.getSelectedZones(),
        format: formatResult,
      });
    });
  });

  const beforePositionExpressionRegex = /^\[/;
  const afterPositionExpressionRegex = /\]$/;

  describe.each([
    ["1", decimalPlaces1, position1],
    ["2", decimalPlaces2, position2],
  ])(
    "currency proposals depend on properties linked to the selected available currency",
    (availableCurrencyIndex, decimalPlaces, positionExpression) => {
      const decimalPlacesRegex = new RegExp("\\.0{" + decimalPlaces + "}");

      test.each([
        ["1", decimalPlacesRegex],
        ["3", decimalPlacesRegex],
        ["5", decimalPlacesRegex],
        ["7", decimalPlacesRegex],
      ])(
        "odd currency proposals depend on decimal places property",
        async (concernedIndex, decimalPlacesRegexp) => {
          await setInputValueAndTrigger(selectors.availableCurrencies, availableCurrencyIndex);
          await setInputValueAndTrigger(selectors.formatProposals, concernedIndex);
          await click(fixture, selectors.applyFormat);
          expect(dispatch).toHaveBeenCalledWith("SET_FORMATTING", {
            sheetId: model.getters.getActiveSheetId(),
            target: model.getters.getSelectedZones(),
            format: expect.stringMatching(decimalPlacesRegexp),
          });
        }
      );

      const firstFourProposalsPositionRegex =
        positionExpression === "after"
          ? afterPositionExpressionRegex
          : beforePositionExpressionRegex;
      const lastFourProposalsPositionRegex =
        positionExpression === "after"
          ? beforePositionExpressionRegex
          : afterPositionExpressionRegex;

      test.each([
        ["0", firstFourProposalsPositionRegex],
        ["1", firstFourProposalsPositionRegex],
        ["2", firstFourProposalsPositionRegex],
        ["3", firstFourProposalsPositionRegex],
        ["4", lastFourProposalsPositionRegex],
        ["5", lastFourProposalsPositionRegex],
        ["6", lastFourProposalsPositionRegex],
        ["7", lastFourProposalsPositionRegex],
      ])(
        "currency proposals depend on position property",
        async (concernedIndex, positionExpressionRegexp) => {
          await setInputValueAndTrigger(selectors.availableCurrencies, availableCurrencyIndex);
          await setInputValueAndTrigger(selectors.formatProposals, concernedIndex);
          await click(fixture, selectors.applyFormat);
          expect(dispatch).toHaveBeenCalledWith("SET_FORMATTING", {
            sheetId: model.getters.getActiveSheetId(),
            target: model.getters.getSelectedZones(),
            format: expect.stringMatching(positionExpressionRegexp),
          });
        }
      );

      const twoDecimalPlacesRegex = /.0{2}/;

      test.each([["1"], ["3"], ["5"], ["7"]])(
        "odd currency proposals have two decimal places when no available currency is selected",
        async (concernedIndex) => {
          await setInputValueAndTrigger(selectors.availableCurrencies, "0");
          await setInputValueAndTrigger(selectors.inputCode, "CODE");
          await setInputValueAndTrigger(selectors.inputSymbol, "SYMBOL");
          await setInputValueAndTrigger(selectors.formatProposals, concernedIndex);
          await click(fixture, selectors.applyFormat);
          expect(dispatch).toHaveBeenCalledWith("SET_FORMATTING", {
            sheetId: model.getters.getActiveSheetId(),
            target: model.getters.getSelectedZones(),
            format: expect.stringMatching(twoDecimalPlacesRegex),
          });
        }
      );

      test.each([
        ["0", afterPositionExpressionRegex],
        ["1", afterPositionExpressionRegex],
        ["2", afterPositionExpressionRegex],
        ["3", afterPositionExpressionRegex],
        ["4", beforePositionExpressionRegex],
        ["5", beforePositionExpressionRegex],
        ["6", beforePositionExpressionRegex],
        ["7", beforePositionExpressionRegex],
      ])(
        "currency first four proposals start by expression placed after digits when no available currency is selected",
        async (concernedIndex, positionExpressionRegexp) => {
          await setInputValueAndTrigger(selectors.availableCurrencies, "0");
          await setInputValueAndTrigger(selectors.inputCode, "CODE");
          await setInputValueAndTrigger(selectors.inputSymbol, "SYMBOL");
          await setInputValueAndTrigger(selectors.formatProposals, concernedIndex);
          await click(fixture, selectors.applyFormat);
          expect(dispatch).toHaveBeenCalledWith("SET_FORMATTING", {
            sheetId: model.getters.getActiveSheetId(),
            target: model.getters.getSelectedZones(),
            format: expect.stringMatching(positionExpressionRegexp),
          });
        }
      );
    }
  );
});

describe("Provided Currencies", () => {
  beforeEach(async () => {
    currenciesContent = Object.assign({}, currenciesRegistry.content);
  });

  afterEach(() => {
    currenciesRegistry.content = currenciesContent;
  });

  test("if currencies are provided in spreadsheet --> display this currencies", async () => {
    const { fixture } = await mountComponent(CustomCurrencyPanel, {
      env: { loadCurrencies },
      model: new Model({}, { external: { loadCurrencies } }),
      props: { onCloseSidePanel: () => {} },
    });
    expect(fixture.querySelector(selectors.availableCurrencies)).toMatchSnapshot();
  });

  test("if currencies aren't provided in spreadsheet --> remove 'available currencies' section", async () => {
    const { fixture } = await mountComponent(CustomCurrencyPanel, {
      env: { loadCurrencies: undefined },
      model: new Model({}),
      props: { onCloseSidePanel: () => {} },
    });
    await nextTick();
    expect(fixture.querySelector(selectors.availableCurrencies)).toBe(null);
  });
});
