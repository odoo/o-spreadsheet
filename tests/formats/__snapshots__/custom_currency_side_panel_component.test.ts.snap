// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provided Currencies if currencies are provided in spreadsheet --> display this currencies 1`] = `
<select
  class="o-input o-available-currencies"
>
  <option
    value="0"
  >
    Custom
  </option>
  <option
    value="1"
  >
    3 decimal places / expression position after (ABC)
  </option>
  <option
    value="2"
  >
    4 decimal places / expression position before (DEF)
  </option>
  
</select>
`;

exports[`custom currency sidePanel component Sidepanel change code input or symbol input --> change currency proposals 1`] = `
[
  "#,##0[$SYMBOL]",
  "#,##0.00[$SYMBOL]",
  "#,##0[$ CODE SYMBOL]",
  "#,##0.00[$ CODE SYMBOL]",
  "[$SYMBOL]#,##0",
  "[$SYMBOL]#,##0.00",
  "[$CODE SYMBOL]#,##0",
  "[$CODE SYMBOL]#,##0.00",
]
`;

exports[`custom currency sidePanel component Sidepanel change code input or symbol input --> change currency proposals 2`] = `
[
  "#,##0[$SYMBOL]",
  "#,##0.00[$SYMBOL]",
  "#,##0[$ TEST SYMBOL]",
  "#,##0.00[$ TEST SYMBOL]",
  "[$SYMBOL]#,##0",
  "[$SYMBOL]#,##0.00",
  "[$TEST SYMBOL]#,##0",
  "[$TEST SYMBOL]#,##0.00",
]
`;

exports[`custom currency sidePanel component Sidepanel change code input or symbol input --> change currency proposals 3`] = `
[
  "#,##0[$SYMBOL]",
  "#,##0.00[$SYMBOL]",
  "#,##0[$ CODE SYMBOL]",
  "#,##0.00[$ CODE SYMBOL]",
  "[$SYMBOL]#,##0",
  "[$SYMBOL]#,##0.00",
  "[$CODE SYMBOL]#,##0",
  "[$CODE SYMBOL]#,##0.00",
]
`;

exports[`custom currency sidePanel component Sidepanel change code input or symbol input --> change currency proposals 4`] = `
[
  "#,##0[$TEST]",
  "#,##0.00[$TEST]",
  "#,##0[$ CODE TEST]",
  "#,##0.00[$ CODE TEST]",
  "[$TEST]#,##0",
  "[$TEST]#,##0.00",
  "[$CODE TEST]#,##0",
  "[$CODE TEST]#,##0.00",
]
`;
